aten/src/ATen/native/cuda/MultinomialKernel.cu

```diff
@@ -228,72 +234,80 @@ sampleMultinomialOnce(int64_t* dest,
       // All threads in bounds load a value
       int cat = chunk * blockDim.x + threadIdx.x;
 
-      accscalar_t a_dist_val = cat < categories ?
-                               static_cast<accscalar_t>(dist[curDist * stride_dist + cat * stride_categories]) / sum :
-                               accZero;
-      scalar_t dist_val = static_cast<scalar_t>(a_dist_val);
+      accscalar_t dist_val = cat < categories ?
+                             static_cast<accscalar_t>(dist[curDist * stride_dist + cat * stride_categories]) / sum :
+                             accZero;
 
-      smem[threadIdx.x] = dist_val;
+      if (is_valid_thread) {
+        smem[threadIdx.x] = dist_val;
+      }
       __syncthreads();
 
       // Perform an inclusive prefix sum of the shared memory contents
-      for (int offset = 1; offset < blockDim.x; offset *= 2) {
-        scalar_t val = zero;
+      for (int offset = 1; offset < valid_block_size; offset *= 2) {
+        accscalar_t val = accZero;
 
-        if (threadIdx.x >= offset) {
+        if (is_valid_thread && threadIdx.x >= offset) {
           val = smem[threadIdx.x - offset] + smem[threadIdx.x];
         }
 
         __syncthreads();
-        if (threadIdx.x >= offset) {
+        if (is_valid_thread && threadIdx.x >= offset) {
           smem[threadIdx.x] = val;
         }
         __syncthreads();
       }
 
-      // Each thread will check to see if the sample falls in its
-      // bucket
-      scalar_t curBucket = smem[threadIdx.x] + prevHighProb;
-      scalar_t prevBucket =
-          threadIdx.x == 0 ? prevHighProb :
-          smem[threadIdx.x - 1] + prevHighProb;
-      bool inBucket =
-          (cat < categories) &&
-          (!(sample >= curBucket) &&
-           (sample >= prevBucket) &&
-           (dist_val > zero));
-
-      if (inBucket) {
-        // We're done; we have the sample
-        // Torch indices are 1-based
-        dest[curDist] = cat;
-        found = true;
-      }
+      if (is_valid_thread) {
+        // Each thread will check to see if the sample falls in its
+        // bucket
+        scalar_t curBucket =
+            static_cast<scalar_t>(smem[threadIdx.x] + prevHighProb);
+        scalar_t prevBucket = static_cast<scalar_t>(
+            threadIdx.x == 0 ? prevHighProb
+                            : smem[threadIdx.x - 1] + prevHighProb);
+        bool inBucket =
+            (cat < categories) &&
+            (!(sample >= curBucket) &&
+            (sample >= prevBucket) &&
+            (dist_val > zero));
+
+        if (inBucket) {
+            // We're done; we have the sample
+            // Torch indices are 1-based
+            atomicMax(&foundPos, cat);
+            found = true;
+        }
 
-      // Store the previous scan's high value for future use
-      prevHighProb = prevHighProb + smem[blockDim.x - 1];
+        // Store the previous scan's high value for future use
+        prevHighProb = prevHighProb + smem[valid_block_size - 1];
```

Comment:
here e.g. valid_block_size is definitely misleading, as the second iteration of this loop can happen only if ncategories > max_block_size, so valid_block_size will be blockDim.x in this case.

Reply:
Ah. You are totally right! The reason why I did it were:

1. `prevHighProb` would be wrong for the next iteration (which, as you said, can't happen)
2. `smem[blockDim.x - 1]` would be out of bounds

I will fix it.